### PR TITLE
merge-homebrew: Allow using git mergetool

### LINKED ISSUE #43: merge-homebrew: Provide way to not open editor
I personally don't like resolving merge conflicts in a standard editor and rely entirely on a dedicated mergetool (like `meld` or `diffuse`).

I don't have time to do this myself at the moment, but as I suspect I'm the only one who cares strongly about this, I'll try to find time to do it when we're done addressing the science migrations.

### PR DESCRIPTION
**Pull request by rwhogg**
Use "git mergetool" if HOMEBREW_MERGETOOL is set.
I'm not a fan of resolving conflicts with a normal
editor.

Closes #43

Signed-off-by: Bob W. Hogg <rwhogg@linux.com>

### CODE REVIEW

**Comment by sjackman** [comment]
Which `mergetool` do you use?

**Comment by rwhogg** [comment]
On Linux, `meld`. On Mac, `tkdiff` because the `meld` cask doesn't like me for some reason.

**Comment by sjackman** [comment]
Thanks for giving me the motivation to finally try out `git mergetool`. The `opendiff` and `FileMerge` of macOS seem not terrible. There's also `gvimdiff`. I'm sure better tools as well.

File: cmd/brew-merge-homebrew.rb [hunk 1]
```diff
@@ -4,6 +4,9 @@
 #:   If `--brew` is passed, merge Homebrew/brew into Linuxbrew/brew.
 #:   If `--core` is passed, merge Homebrew/homebrew-core into Linuxbrew/homebrew-core.
 #:   If <commit> is passed, merge only up to that upstream SHA-1 commit.
+#:
+#:   If HOMEBREW_MERGETOOL is set, run "git mergetool" instead of opening conflicting
+#:   files in an editor.
```
Comment (sjackman) [inline cmd/brew-merge-homebrew.rb]: Please remove this chunk.

File: cmd/brew-merge-homebrew.rb [hunk 2]
```diff
@@ -48,7 +51,11 @@ def resolve_conflicts
     return conflicts if conflicts.empty?
     oh1 "Conflicts"
     puts conflicts.join(" ")
-    safe_system *editor, *conflicts
+    unless ENV["HOMEBREW_MERGETOOL"].nil?
```
Comment (sjackman) [inline cmd/brew-merge-homebrew.rb]: Please change this to something along the lines of
```ruby
if system "git config merge.tool >/dev/null 2>/dev/null"
```
and cache the result.

**Comment by iMichka** [comment]
I like to set `merge.conflictstyle diff3` when looking at merge diffs. Tried that with linuxbrew, it breaks at multiple places, so I gave up. But this is what I use at work + vim.

**Comment by rwhogg** [comment]
Edited; thanks for the suggestions Shaun!

File: cmd/brew-merge-homebrew.rb [hunk 3]
```diff
@@ -21,6 +21,10 @@ def git
     @git ||= Utils.git_path
   end
 
+  def has_mergetool
+    @mergetool ||= system "git config merge.tool >/dev/null 2>/dev/null"
```
Comment (sjackman) [inline cmd/brew-merge-homebrew.rb]: The `||=` trick doesn't work well for Boolean values. If the value is false, the evaluation of the right-hand-side won't be skipped next time, as you'd like. To memoize a boolean, you can use…
```ruby
@mergetool = system "git config merge.tool >/dev/null 2>/dev/null" if @mergetool.nil?
```
See https://stackoverflow.com/questions/11158972/how-can-i-memoize-a-method-that-may-return-true-false-or-nil-in-ruby

File: cmd/brew-merge-homebrew.rb [hunk 4]
```diff
@@ -21,6 +21,10 @@ def git
     @git ||= Utils.git_path
   end
 
+  def has_mergetool
```
Comment (sjackman) [inline cmd/brew-merge-homebrew.rb]: Ruby style is that predicates (functions that return bool) should be named with a  question mark, like `def has_mergetool?`

Comment (rwhogg) [reply cmd/brew-merge-homebrew.rb]: D'oh, forgot to do that! Thanks.

**Comment by rwhogg** [comment]
Thanks @sjackman, think I've finally got it this time.

**Pull request merged**